### ISSUE TITLE
ccls does not recognize references in struct initializers inside arrays

### ISSUE THREAD

**Issue by cmm** [MaskRay/ccls#76]
`ccls` version: 0.20180913
editor: `Emacs`, with latest Melpa versions of `lsp-mode`, `ccls` etc as of 2018-09-18

This is a minimal example:
     int x;
     struct { int *v; } arr[] = {
                                { .v = &x; },
                                {  NULL },
     };

And its accompanying `compile_commands.json` (assuming the C file is called `foo.c`):
    [{"file":"foo.c","command":"gcc -O0 -g -c foo.c -o foo.o"}]

The problem is that `ccls` does not consider the `&x` to be a reference to `x`.  What seems to be throwing it off is the `{ NULL }` element at the end of the array: if I comment it out entirely or change it to just `{}`, the problem goes away.

(I realize that the `{ NULL }` thing looks slightly weird, but the problem occurs on a large code base with lots of instances of this construct, so changing the code to accommodate `ccls` is not really an option.  Also I think `rtags` (which I'm trying to migrate from because it's too slow...) did not have the problem).

**Comment by cmm** [comment]
For the record: the _actual_ problem turned out to be that ccls failed to find some include files (my fault: I actually build stuff in a Docker container, but running ccls there is not a good option for several boring reasons; the things that were failing to be indexed were cross-compiled to x86 with -freestanding but with compiler include path given explicitly -- i.e. the missing headers were "stdarg.h" and such, because I didn't have the same exact cross-compiler version installed on the indexing machine...), and apparently ccls is really fragile about that.

Now, such fragility in itself is not necessarily a problem (although neither CQuery nor RTags seem to share it), but failure to report the problem clearly (or, in fact, at all -- at least in a way I could detect it without guesswork) together with resulting failure to index seemingly-unrelated stuff is a serious problem, IMHO.  Wish I had time to help here...

**Comment by MaskRay** [comment]
>  ccls failed to find some include files (my fault: I actually build stuff in a Docker container

I don't think it is an issue of ccls. ccls does very little to the compiler command line. It merely adds `-resource-dir` -working-directory`. The heavylifting work is done by clangDriver. clang is inherently different from GCC, while it emulates it quiet well, you may need to do more to emulate a cross GCC.

You may experiment with:

`fclang++ -fsyntax-only -Xclang -code-completion-at=a.cc:4:1 a.cc <other options>`

to figure out a usable compiler command line.

For cross compiling, this command will help you figure out if clangDriver detects the correct system search directories. If not, add some `-isystem`:

`clang++ -target aarch64 -v -E -xc++ /dev/null`

I think cquery and rtags do more on twiddling command line options. Some changes are necessary because they are libclang (Clang C) based and libclang does not provide an interface for these clangDriver stuff.

**Comment by Riatre** [comment]
I think what @cmm suggests here is to warn the user "there is something wrong" in your includes, which has the potential of helping in more than just troubleshooting include path issues.

What happened here in his case is `stdarg.h` is missing, and this was reported somewhere in diagnostics, but not in his `foo.c`. For example, imagine that you have three files:

a.c
```c
#include "b.h"

// do something
```

b.h
```c
#pragma once
#include "c.h"

// do something
```

c.h
```c
#pragma once
#include <something_that_does_not_exists.h>
```

Currently, while generating diagnostics, ccls does not propagate the errors in header files into the file that includes it. So no diagnostics reported for `a.c` about missing `something_that_does_not_exists.h`.

Most of time you will notice that there are some "undefined reference" errors reported by diagnostics in `a.c`, then you may guess what happened and try to diagnose. But sometimes the error could be really funny, for example when there are `#include "xxx.inc"`s, `xxx.inc` has `{}`s which is expected to be matched in the outer file and due to some reason `xxx.inc` could not be found. It is simply not possible to recover from these errors but clang still tries VERY hard to do so while generating diagnostics, lead to bogus "undefined xxx" everywhere.

If adding an option to propagate errors in headers into the parent file sounds interesting to you, I may check whether it is possible to do so in the next few days.

**Comment by MaskRay** [comment]
@Riatre Propagating errors in headers to the main file sounds good. Appreciated if you'd like to contribute!

I think you'll probably need to look into `concerned` (which I named `inside_main` before)

* `inside_main`: I used `SourceManager::isInMainFile` from `clang/Basic/SourceManager.h` to check if the diagnostic was in the main file.
* `concerned`: the rename was because I thought a header file (`a.h`) should not be parsed by itself, but rather through its import file (`a.c`). By doing that, `isInMainFile` was no longer appropriate.
* In commit 84984c6c2720df8454b4c0bfec296db792253792, I changed back to parse the header directly (added a new parameter `can_be_inferred` to `Project::FindEntry`). By parsing the import file, preamble has to be disabled to take into account of update of `a.h`, this turns out to be too slow for some clang `.h files.

Hope these will help you understand the mechanism a bit.

**Comment by cmm** [comment]
Come to think about it, another problem here is that compiler-specific "magic" standard includes (like "stdarg.h" or "stdbool.h") are not recognized by ccls as such (or maybe the other indexers just simply have a fallback path to look for those?).  It's of course cool how in my case ccls finds the "stdarg.h" belonging to the specific ancient version of cross-gcc and succeeds in parsing & indexing it, but that success strikes me as very lucky, and not something one should reasonably expect to always work.

The right thing would be to find and index clang's "stdarg.h" & friends (of the specific clang version that ccls is built upon, that is).

**Comment by MaskRay** [comment]
`std{arg,bool}.h` are included in the resource directory shipped with clang.

```
# Release/lib/clang/8.0.0 /usr/lib/clang/3.9 are examples of -resource-dir
% ls Release/lib/clang/8.0.0/include/std{arg,bool}.h 
Release/lib/clang/8.0.0/include/stdarg.h  Release/lib/clang/8.0.0/include/stdbool.h
% ls /usr/lib/clang/3.9/include/std{arg,bool}.h
/usr/lib/clang/3.9/include/stdarg.h  /usr/lib/clang/3.9/include/stdbool.h
```

You may check https://github.com/MaskRay/ccls/wiki/FAQ#check-if--resource-dir-is-correct if your resource directory is correctly configured in your container and host.

**Comment by cmm** [comment]
Hrm, I swear it was finding the ones under /usr/lib/gcc/x86_64-linux-gnu/5/include/ for me before, but it does indeed find the clang ones now.  Maybe something to do with `:index (:multiVersion 1)`...

**Comment by MaskRay** [comment]
> Hrm, I swear it was finding the ones under /usr/lib/gcc/x86_64-linux-gnu/5/include/ for me before, but it does indeed find the clang ones now. Maybe something to do with `:index (:multiVersion 1)`...

You may want to check the logic in https://github.com/llvm-mirror/clang/tree/master/lib/Frontend/InitHeaderSearch.cpp#L241

`index.multiVersion: 1` is still a highly experimental feature (that I don't use daily). The cache files will be bloated a lot. That is why I added `index.multiVersionBlacklist`

**Comment by amosbird** [comment]
>  I may check whether it is possible to do so in the next few days.

@Riatre Great. This may also help debugging other weird diagnostic issues. How about making this `error propagation ` flavored diagnostic a standalone command? From what I can tell, it's main purpose isn't to aid daily developing process, but for diagnosing clang-based tools' configuration errors. It also keeps the original diag thread clean and robust.
